server/plugin_release.go

```diff
@@ -29,199 +35,406 @@ import (
+	s3PluginSignatureFilepath := filepath.Join(tmpFolder, fmt.Sprintf("%v-%v.tar.gz.sig", repositoryName, tag))
+	if err := os.Symlink(githubPluginSignatureFilePath, s3PluginSignatureFilepath); err != nil {
+		return fmt.Errorf("failed to duplicate signature file err=%w", err)
+	}
+
+	s3Bucket := []string{s3PluginSignatureFilepath}
+	for _, p := range platformPluginFilePaths {
+		s3Bucket = append(s3Bucket, p)
+		s3Bucket = append(s3Bucket, fmt.Sprintf("%s.sig", p))
+	}
+
+	// Upload plugins and signatures to s3 release bucket
+	if err := uploadToS3(ctx, Cfg, s3Bucket); err != nil {
+		return fmt.Errorf("failed to upload to s3 err=%w", err)
+	}
+
+	return nil
+}
+
+// checkRepo checks if repo exists.
+func checkRepo(cfg *MatterbuildConfig, owner, repo string) error {
+	ctx := context.Background()
+	ts := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: cfg.GithubAccessToken})
+	tc := oauth2.NewClient(ctx, ts)
+	client := github.NewClient(tc)
+
+	if _, _, err := client.Repositories.ListBranches(ctx, owner, repo, nil); err != nil {
 		LogError("No branch found. err=" + err.Error())
 		return fmt.Errorf("Looks like this Repository is not part of the org or does not exist. Repo: %s", repo)
-
 	}
 
 	return nil
 }
 
-func createTag(tag, repo string) error {
-	ts := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: Cfg.GithubAccessToken})
+// getReleaseByTag gets github release by tag.
+func getReleaseByTag(cfg *MatterbuildConfig, owner, repositoryName, tag string) (*github.RepositoryRelease, error) {
+	ctx := context.Background()
+	ts := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: cfg.GithubAccessToken})
 	tc := oauth2.NewClient(ctx, ts)
-	client = github.NewClient(tc)
+	client := github.NewClient(tc)
 
-	refs, _, err := client.Git.ListRefs(ctx, Cfg.GithubOrg, repo, nil)
+	release, _, err := client.Repositories.GetReleaseByTag(ctx, owner, repositoryName, tag)
 	if err != nil {
-		return err
+		return nil, fmt.Errorf("failed to get release by tag err=%w", err)
 	}
-	checkTag := fmt.Sprintf("refs/tags/%s", tag)
-	for _, ref := range refs {
-		if ref.GetRef() == checkTag {
-			return fmt.Errorf("Tag %s already exist. Please use another one", tag)
+
+	return release, nil
+}
+
+// createTag creates a new tag at master for repo.
+// Returns ErrTagExists if tag already exists, nil if successful and an error otherwise.
+func createTag(cfg *MatterbuildConfig, owner, tag, repository string) error {
+	ctx := context.Background()
+	ts := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: cfg.GithubAccessToken})
+	tc := oauth2.NewClient(ctx, ts)
+	client := github.NewClient(tc)
+
+	refs, _, err := client.Git.GetRefs(ctx, owner, repository, fmt.Sprintf("tags/%s", tag))
+	if err != nil {
+		if err, ok := err.(*github.ErrorResponse); ok && err.Response.StatusCode == http.StatusNotFound {
+			LogInfo("tag %s was not found, moving on to creating tag", tag)
+		} else {
+			return err
 		}
+	} else if len(refs) > 0 {
```

Comment:
because i only want to check `refs` iff `err == nil`, if i drop the `else` `len(refs)` will always be check even if `err != nil`. Normally this wouldn't be an issue but we have a case where we fall through and ignore `err` value when `err == NotFound`